printer/src/printer.rs

```diff
@@ -0,0 +1,334 @@
+  }
+
+  fn print_text(&mut self, text: &Text) {
+    self.push_str(text.content);
+  }
+
+  fn print_escape(&mut self, escape: &Escape) {
+    self.push('\\');
+    self.push(escape.escaped_char);
+  }
+
+  fn print_expression(&mut self, expr: &Expression) {
+    dispatch_print!(self, expr, Expression {
+      AnnotationExpression => print_annotation_expression,
+      LiteralExpression => print_literal_expression,
+      VariableExpression => print_variable_expression,
+    });
+  }
+
+  fn print_annotation_expression(&mut self, expr: &AnnotationExpression) {
+    self.helper_print_expression(
+      None::<()>,
+      Some(&expr.annotation),
+      &expr.attributes,
+      |_, _| {},
+    );
+  }
+
+  fn print_literal_expression(&mut self, expr: &LiteralExpression) {
+    self.helper_print_expression(
+      &expr.literal,
+      expr.annotation.as_ref(),
+      &expr.attributes,
+      Self::print_literal,
+    );
+  }
+
+  fn print_variable_expression(&mut self, expr: &VariableExpression) {
+    self.helper_print_expression(
+      &expr.variable,
+      expr.annotation.as_ref(),
+      &expr.attributes,
+      Self::print_variable,
+    );
+  }
+
+  fn helper_print_expression<T, F>(
+    &mut self,
+    body: T,
+    annotation: Option<&Annotation>,
+    attributes: &Vec<Attribute>,
+    cb: F,
+  ) where
+    F: FnOnce(&mut Self, T),
+  {
+    self.push('{');
+    self.push(' ');
+
+    cb(self, body);
+
+    if let Some(annotation) = annotation {
+      if !matches!(self.out.chars().last(), Some(' ')) {
+        self.push(' ');
+      }
+
+      let Annotation::Function(fun) = annotation;
+      self.print_function(fun);
+    }
+
+    for attr in attributes {
+      self.print_attribute(attr);
+    }
+
+    self.push(' ');
+    self.push('}');
+  }
+
+  fn print_function(&mut self, fun: &Function) {
+    self.push(':');
+    self.print_identifier(&fun.id);
+
+    for option in &fun.options {
+      self.print_option(option);
+    }
+  }
+
+  fn print_identifier(&mut self, id: &Identifier) {
+    if let Some(namespace) = id.namespace {
+      self.push_str(namespace);
+      self.push(':');
+    }
+    self.push_str(id.name);
+  }
+
+  fn print_option(&mut self, option: &FnOrMarkupOption) {
+    self.push(' ');
+    self.print_identifier(&option.key);
+    self.push('=');
+    dispatch_print!(self, &option.value, LiteralOrVariable {
+      Literal => print_literal,
+      Variable => print_variable,
+    })
+  }
+
+  fn print_literal(&mut self, lit: &Literal) {
+    dispatch_print!(self, lit, Literal {
+      Text => print_text,
+      Quoted => print_quoted,
+      Number => print_number,
+    });
+  }
+
+  fn print_quoted(&mut self, quoted: &Quoted) {
+    self.push('|');
+    for part in &quoted.parts {
+      dispatch_print!(self, part, QuotedPart {
+        Text => print_text,
+        Escape => print_escape,
+      });
+    }
+    self.push('|');
+  }
+
+  fn print_number(&mut self, num: &Number) {
+    self.push_str(num.raw);
```

Comment:
I don't think we should format invalid code to valid code, as it is always ambiguous (should we add a `0`, or remove the `.`?). This can be done through autofixes that users have to explicitly opt into.